[Data] Fix pandas memory calculation.

Bye-legumes opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

close https://github.com/ray-project/ray/issues/46785
Current the memory usage for pandas is not accurate when it's object, so we just implement to calculated it in recursion in case of nested.
## Related issue number

closes https://github.com/ray-project/ray/issues/46785, closes https://github.com/ray-project/ray/issues/48506

## Checks

- [√] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [√] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [√] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

Bye-legumes commented:
@bveeramani  @c21 can you check this PR. I think the main idea is to calculate it in recursion.. I just did some test and for the example in https://github.com/ray-project/ray/issues/46785, each time the time is 20ms for calculation, I think it's acceptable?

Bye-legumes commented:
Although it may not support tensor or other object.

Bye-legumes commented:
hi, @bveeramani can you check this?

richardliaw commented:
Hi @Bye-legumes , I can take over the review here. 

I actually implemented a couple tests, but seems like this is failing here:

```python
___________________________ test_size_bytes_bytes_object ____________________________

ray_start_regular_shared = RayContext(dashboard_url='127.0.0.1:8265', python_version='3.10.15', ray_version='3.0.0.dev0', ray_commit='9e5c5bfba9d79d1d134719cf202c7150407acbc6')

    def test_size_bytes_bytes_object(ray_start_regular_shared):
        def generate_data(batch):
            for _ in range(8):
                yield {"data": [[b"\x00" * 128 * 1024 * 128]]}
    
        ds = (
            ray.data.range(1, override_num_blocks=1)
            .map_batches(generate_data, batch_size=1)
            .map_batches(lambda batch: batch, batch_format="pandas")
        )
    
        true_value = 128 * 1024 * 128 * 8
        for bundle in ds.iter_internal_ref_bundles():
            size = bundle.size_bytes()
            # assert that true_value is within 10% of bundle.size_bytes()
>           assert true_value * 0.9 <= size <= true_value * 1.1, (true_value, size)
E           AssertionError: (134217728, 192)
E           assert (134217728 * 0.9) <= 192
```

richardliaw commented:
A couple tests I wrote --

```

def test_size_bytes_small(ray_start_regular_shared):
    animals = ["Flamingo", "Centipede"]
    block = pd.DataFrame({"animals": animals})
    block["animals"] = block["animals"].astype("string")

    block_accessor = PandasBlockAccessor.for_block(block)
    bytes_size = block_accessor.size_bytes()

    # generally strings are hard, so let's use what Pandas gives us.
    # get memory usage from pandas
    memory_usage = block.memory_usage(index=True, deep=True).sum()
    # check that memory usage is within 10% of the size_bytes
    assert memory_usage * 0.9 <= bytes_size <= memory_usage * 1.1, (
        bytes_size,
        memory_usage,
    )


def test_size_bytes_large_str(ray_start_regular_shared):
    animals = [
        random.choice(["alligator", "crocodile", "centipede", "flamingo"])
        for i in range(100_000)
    ]
    block = pd.DataFrame({"animals": animals})
    block["animals"] = block["animals"].astype("string")

    block_accessor = PandasBlockAccessor.for_block(block)
    bytes_size = block_accessor.size_bytes()

    # String disk usage is wildly different from in-process memory usage
    memory_usage = block.memory_usage(index=True, deep=True).sum()
    # check that memory usage is within 10% of the size_bytes
    assert memory_usage * 0.9 <= bytes_size <= memory_usage * 1.1, (
        bytes_size,
        memory_usage,
    )


def test_size_bytes_large_floats(ray_start_regular_shared):
    animals = [random.random() for i in range(100_000)]
    block = pd.DataFrame({"animals": animals})

    block_accessor = PandasBlockAccessor.for_block(block)
    bytes_size = block_accessor.size_bytes()

    memory_usage = pickle.dumps(block).__sizeof__()
    # check that memory usage is within 10% of the size_bytes
    assert memory_usage * 0.9 <= bytes_size <= memory_usage * 1.1, (
        bytes_size,
        memory_usage,
    )


def test_size_bytes_bytes_object(ray_start_regular_shared):
    def generate_data(batch):
        for _ in range(8):
            yield {"data": [[b"\x00" * 128 * 1024 * 128]]}

    ds = (
        ray.data.range(1, override_num_blocks=1)
        .map_batches(generate_data, batch_size=1)
        .map_batches(lambda batch: batch, batch_format="pandas")
    )

    true_value = 128 * 1024 * 128 * 8
    for bundle in ds.iter_internal_ref_bundles():
        size = bundle.size_bytes()
        # assert that true_value is within 10% of bundle.size_bytes()
        assert true_value * 0.9 <= size <= true_value * 1.1, (true_value, size)


def test_size_bytes_unowned_numpy(ray_start_regular_shared):
    import numpy as np

    df = pd.DataFrame(
        {
            "data": [
                np.random.randint(size=1024, low=0, high=100, dtype=np.int8)
                for _ in range(1_000)
            ],
        }
    )

    block_accessor = PandasBlockAccessor.for_block(df)
    block_size = block_accessor.size_bytes()
    true_value = 1024 * 1000
    assert true_value * 0.9 <= block_size <= true_value * 1.1
```

Bye-legumes commented:
> Hi @Bye-legumes , I can take over the review here.
> 
> I actually implemented a couple tests, but seems like this is failing here:
> 
> ```python
> ___________________________ test_size_bytes_bytes_object ____________________________
> 
> ray_start_regular_shared = RayContext(dashboard_url='127.0.0.1:8265', python_version='3.10.15', ray_version='3.0.0.dev0', ray_commit='9e5c5bfba9d79d1d134719cf202c7150407acbc6')
> 
>     def test_size_bytes_bytes_object(ray_start_regular_shared):
>         def generate_data(batch):
>             for _ in range(8):
>                 yield {"data": [[b"\x00" * 128 * 1024 * 128]]}
>     
>         ds = (
>             ray.data.range(1, override_num_blocks=1)
>             .map_batches(generate_data, batch_size=1)
>             .map_batches(lambda batch: batch, batch_format="pandas")
>         )
>     
>         true_value = 128 * 1024 * 128 * 8
>         for bundle in ds.iter_internal_ref_bundles():
>             size = bundle.size_bytes()
>             # assert that true_value is within 10% of bundle.size_bytes()
> >           assert true_value * 0.9 <= size <= true_value * 1.1, (true_value, size)
> E           AssertionError: (134217728, 192)
> E           assert (134217728 * 0.9) <= 192
> ```

Thanks so much! I think this time it's OK now!

richardliaw commented:
Hi, awesome - could you make sure to add a couple tests to test_pandas_block?

Bye-legumes commented:
> Hi, awesome - could you make sure to add a couple tests to test_pandas_block?

HI, I just add some test for test_pandas_block, can you check them? Thx!

richardliaw commented:
Yes, taking a look - I hope you don't mind if I push to your PR?

Bye-legumes commented:
> Yes, taking a look - I hope you don't mind if I push to your PR?

Sure! just push the codes!

richardliaw commented on python/ray/data/tests/test_pandas_block.py:
```diff
@@ -48,7 +54,203 @@ def fn2(batch):
+            memory_usage,
+        )
+
+    def test_size_bytes_large_str(ray_start_regular_shared):
+        animals = [
+            random.choice(["alligator", "crocodile", "centipede", "flamingo"])
+            for i in range(100_000)
+        ]
+        block = pd.DataFrame({"animals": animals})
+        # if you remove this line it breaks
+        block["animals"] = block["animals"].astype("string")
```
@Bye-legumes could you take a look at this?

Bye-legumes replied:
For this test, if we use ` block["animals"] = block["animals"].astype("string")`, the detype is `animals    string[python]
dtype: object` and we can use `memory_usage = block.memory_usage(index=True, deep=True).sum()` to get the correct memory.
if we do not use 
` block["animals"] = block["animals"].astype("string")`, it will be the object `animals    object
dtype: object
5775025`, and  `memory_usage = block.memory_usage(index=True, deep=True).sum()` cannot give the correct usage as this issue motioned before and the correct bytes is 58*100_000 in real. the real size is the size of string in list + list ref size `memory_usage = sys.getsizeof(animals[0])*100_000
`

richardliaw commented on python/ray/data/tests/test_pandas_block.py:
```diff
@@ -48,7 +54,203 @@ def fn2(batch):
+        true_size = rows * list_overhead
+        assert true_size * 0.9 <= bytes_size <= true_size * 1.1, (bytes_size, true_size)
+
+    def test_size_bytes_multi_level_nesting(ray_start_regular_shared):
+        rows = 1_000
+        size = 10  # if you change this to 1024 it breaks
+        data = {
+            "complex": [
+                {"list": [np.random.rand(size)], "value": {"key": "val"}}
+                for _ in range(rows)
+            ],
+        }
+        block = pd.DataFrame(data)
+        block_accessor = PandasBlockAccessor.for_block(block)
+        bytes_size = block_accessor.size_bytes()
+
+        # Manually calculate the size
+        list_overhead = sys.getsizeof([0] * size) + size * 28
```
@Bye-legumes could you take a look at this? if you change size to 1024, it breaks. would be great to get this calculation right

Bye-legumes replied:
I just resolve the calculation of the size which count the string.

richardliaw reviewed (commented):
I updated the tests for readability and also made it a bit more extensive -- I found some areas that broke, could you take a look?

Bye-legumes commented:
> I updated the tests for readability and also made it a bit more extensive -- I found some areas that broke, could you take a look?

Hi, I just fixed the issues that you mentioned! Thx!

richardliaw reviewed (approved):
This looks great to me!

Bye-legumes commented:
> This looks great to me! Will ping others once tests pass.

Hi! I think it's OK now!

raulchen commented on python/ray/data/_internal/pandas_block.py:
```diff
@@ -294,7 +295,65 @@ def num_rows(self) -> int:
+        object_need_check = ["object", "python_object()"]
+        # Handle object columns separately
+        for column in self._table.columns:
+            if str(self._table[column].dtype) in object_need_check:
```
nit, can we directly compare the dtype without casting it to strings?

Bye-legumes replied:
That is my original solution but if will fail in some case as the `self._table[column].dtype` will return python_object(). It will happens when there is a nested dict. I cannot find the type called "python_object()" so I just check it manually with String.

raulchen commented on python/ray/data/tests/test_pandas_block.py:
```diff
@@ -48,7 +54,233 @@ def fn2(batch):
+        # check that memory usage is within 10% of the size_bytes
+        # For strings, Pandas seems to be fairly accurate, so let's use that.
+        memory_usage = block.memory_usage(index=True, deep=True).sum()
+        assert memory_usage * 0.9 <= bytes_size <= memory_usage * 1.1, (
```
nit, can use `from pytest import approx`

Bye-legumes replied:
fixed!

raulchen commented on python/ray/data/tests/test_pandas_block.py:
```diff
@@ -48,7 +54,233 @@ def fn2(batch):
+                ]
+            )
+            / 4
+        )
```
what is the purpose of testing the size against mean_size * num? 
Why not just simply test against the real size, like the above test case does?

raulchen commented on python/ray/data/tests/test_pandas_block.py:
```diff
@@ -48,7 +54,233 @@ def fn2(batch):
+            # assert that true_value is within 10% of bundle.size_bytes()
+            assert true_value * 0.9 <= size <= true_value * 1.1, (true_value, size)
+
+    def test_unowned_numpy(ray_start_regular_shared):
```
for my own understanding, what does unowned mean?

Bye-legumes replied:
Changed to a better name "nested", as it's nd.arrary in list.

raulchen commented on python/ray/data/tests/test_pandas_block.py:
```diff
@@ -48,7 +54,233 @@ def fn2(batch):
+        int_size = rows * 8
+        float_size = rows * 8
+        str_size = (
+            rows * sum([sys.getsizeof(s) for s in ["apple", "banana", "cherry"]]) // 3
```
same question as the above on why testing against the average size. 
This seems fragile if every item happens to be the longest/shortest string. 
I don't think `random.choice` is necessary here. May be just a fixed number for each fruit type?

Bye-legumes replied:
Just fixed with checking the real data. And fixed some tests so if it use random, it will calculated the real size instead of mean/sum

richardliaw commented:
Some tests failing?

Bye-legumes commented:
> Some tests failing?

Yeah, just fix as the some of the inner blocks are TensorArrayElement, TensorDtype. I just make some changes and also I can use type check now. In the previous it's just "python_object()" from extension array.

Bye-legumes commented:
> Some tests failing?

All the test are OK now.

richardliaw commented:
Awesome, thank you so much @Bye-legumes !!!!

aslonnie commented:
This PR is reverted. many data release tests are broken.
